clang/lib/Sema/SemaStmt.cpp

```diff
@@ -4418,3 +4418,57 @@ StmtResult Sema::ActOnCapturedRegionEnd(Stmt *S) {
 
   return Res;
 }
+
+WhereClause *Sema::ActOnWhereClause(SourceLocation Loc) {
+  return new (Context) WhereClause(Loc);
+}
+
+BoundsFact *Sema::ActOnBoundsFact(IdentifierInfo *Id, Expr *E,
+                                  Scope *CurScope, SourceLocation Loc) {
+  BoundsExpr *Bounds = dyn_cast<BoundsExpr>(E);
+  if (!Bounds)
+    return nullptr;
+
+  LookupResult Lookup(*this, Id, Loc, Sema::LookupOrdinaryName);
+  LookupParsedName(Lookup, CurScope, nullptr, true);
+  if (Lookup.empty()) {
+    Diag(Loc, diag::err_undeclared_var_use) << Id->getName();
+    return nullptr;
+  }
+
+  VarDecl *VD = Lookup.getAsSingle<VarDecl>();
+  if (!VD)
+    return nullptr;
+
+  return new (Context) BoundsFact(VD, Bounds, Loc);
+}
+
+EqualityOpFact *Sema::ActOnEqualityOpFact(Expr *E, SourceLocation Loc) {
+  // We define an equality-op fact in terms of equality-expressions as defined
```

Comment:
It would be helpful to add a comment "We are checking whether E is an equality expression".